### PR TITLE
Feature genome browser (part 1)

### PR DESCRIPTION
**Pull request by JenkoB**
This is the first PR with changes on processors to support Genome Browsers.
Here are changes in 
  - upload-genome processor
  - processors for annotation files (upload-gtf, upload-gff3, gff-to-gtf, upload-bed)
 - chip-seq processors (macs14, macs2)

Next part will be adding BigWig output to processors with aligners.

@mstajdohar @jkokosar @JureZmrzlikar please review.

### CODE REVIEW

File: resolwe_bio/processes/chip_seq/rose2.yml [hunk 1]
```diff
@@ -110,30 +116,59 @@
...
+      {% endif %}
+
       {% if input_upload %}
-        NAME=`basename {{ input_upload.bed.file }} .bed`
+        if [[ {{ input_upload.bed.file }} == *.bed ]]; then
+          NAME=`basename {{ input_upload.bed.file }} .bed`
+          cp {{ input_upload.bed.file }} "${NAME}.bed"
+        else
+          NAME=`basename {{ input.peaks_bed.file }} .bed.gz`
+          bgzip -cd {{ input.peaks_bed.file }} > "${NAME}.bed"
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/chip_seq/rose2.yml]: Just a comment: you can also use `zcat`

File: resolwe_bio/processes/import_data/genome.yml [hunk 2]
```diff
@@ -114,6 +114,10 @@
     - name: dict
       label: Fasta dict
       type: basic:file
+    - name: fasta_track_jbrowse
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/import_data/genome.yml]: Bump process version.

File: resolwe_bio/tests/processes/test_support_processors.py [hunk 3]
```diff
@@ -87,6 +87,9 @@ def test_gff_to_gtf(self):
 
         gff_to_gtf = self.run_process('gff-to-gtf', {'annotation': annotation.id})
         self.assertFile(gff_to_gtf, 'annot', 'gff_to_gtf_annotation.gtf')
+        del gff_to_gtf.output['annot_sorted_track_jbrowse']['total_size']  # Non-deterministic output.
+        self.assertFields(gff_to_gtf, 'annot_sorted_track_jbrowse', {'refs': ['tracks/annotation'],
+                                                            'file': 'trackList.json'})
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_support_processors.py]: This line wrapping does not seem right. Also check similar examples below.

File: resolwe_bio/tests/processes/test_upload.py [hunk 4]
```diff
@@ -413,3 +417,36 @@ def test_upload_vcf(self):
...
         self.assertFields(vcf, 'species', 'Homo sapiens')
         self.assertFields(vcf, 'build', 'b37')
+
+    @tag_process('upload-gff3')
+    def test_upload_gff3(self):
+        inputs = {
+            'src': 'PGSC upload.gff3',
+            'build': 'ST',
+            'source': 'PGSC',
+            'species': 'Solanum tuberosum'
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_upload.py]: Add comma at the end of line. Same applies to the test below.

File: resolwe_bio/tools/bed_file_corrections_genome_browsers.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python3
+"""Maximize 5th column of bed file to 1000."""
```
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bed_file_corrections_genome_browsers.py]: I would rephrase this. To me, "maximize" means that something is **raised** (to the full potential). But if i see correctly, you are actually **lowering** the scores to be below 1000.  I would rather see something like "Restrict all scores to the maximal value of `LIMIT`." And set a constant `LIMIT=1000`.

File: resolwe_bio/processes/chip_seq/rose2.yml [hunk 6]
```diff
@@ -98,6 +98,12 @@
     - name: scatter_plot
       label: Super-Enhancer plot
       type: basic:json
+    - name: species
```
[the remark this answers is not in the document]
Comment (JenkoB) [reply resolwe_bio/processes/chip_seq/rose2.yml]: Good point! I haven't thought on that.

File: resolwe_bio/processes/import_data/annotation.yml [hunk 7]
```diff
@@ -65,7 +65,16 @@
       type: basic:string
   output:
     - name: annot
-      label: Uploaded file
+      label: Uploaded GFF3 file
+      type: basic:file
+    - name: annot_sorted
+      label: Sotred GFF3 file
```
Comment (jkokosar) [inline resolwe_bio/processes/import_data/annotation.yml]: Sorted

File: resolwe_bio/processes/chip_seq/macs14.yml [hunk 8]
```diff
@@ -38,14 +38,34 @@
...
+      type: basic:file
     - name: peaks_xls
       label: Peaks (XLS)
       type: basic:file
     - name: wiggle
       label: Wiggle
       type: basic:file
-    - name: summits_bed
-      label: Summits (BED)
+    - name: control_bigwig_all_gbrowsers
```
Comment (jkokosar) [inline resolwe_bio/processes/chip_seq/macs14.yml]: This should be `control_bigwig` ( and `treat_bigwig` bellow) to keep the name more consise

File: resolwe_bio/processes/chip_seq/macs14.yml [hunk 9]
```diff
@@ -38,14 +38,34 @@
...
+      label: Control (bigWig)
+      type: basic:file
+      required: false
+    - name: treat_bigwig_all_gbrowsers
+      label: Treat (bigWig)
+      type: basic:file
+      required: false
```
[the remark this answers is not in the document]
Comment (JenkoB) [reply resolwe_bio/processes/chip_seq/macs14.yml]: Thanks. It is required.

File: resolwe_bio/processes/import_data/annotation.yml [hunk 10]
```diff
@@ -9,11 +9,11 @@
     expression-engine: jinja
     executor:
       docker:
-        image: resolwebio/legacy:1.0.0
+        image: resolwebio/rnaseq:3.2.0
```
Comment (mstajdohar) [inline resolwe_bio/processes/import_data/annotation.yml]: Nice!

File: resolwe_bio/processes/import_data/annotation.yml [hunk 11]
```diff
@@ -171,7 +204,21 @@
...
+      igvtools index "${NAME}"_sorted.gtf || re-warning "Indexing of GTF file failed."
+
+      flatfile-to-json.pl --gff "${NAME}"_sorted.gtf --out . --trackLabel "annotation"
+      re-checkrc "Annotation track processing for JBrowse failed."
+
       re-save-file annot "${NAME}.gtf"
+      re-save-file annot_sorted "${NAME}"_sorted.gtf
+      re-save-file annot_sorted_track_jbrowse trackList.json tracks/annotation
+
+      if [ -f "${NAME}"_sorted.gtf.idx ]; then
```
Comment (mstajdohar) [inline resolwe_bio/processes/import_data/annotation.yml]: Did you write a test for this case?

File: docs/CHANGELOG.rst [hunk 12]
```diff
@@ -97,6 +98,7 @@ Fixed
 
   - ``upload-gff3``
   - ``upload-gtf``
+  - ``gff-to-gtf``
```
Comment (mstajdohar) [inline docs/CHANGELOG.rst]: Duplicated.

File: resolwe_bio/processes/support_processors/gff_to_gtf.yml [hunk 13]
```diff
@@ -10,23 +10,31 @@
     expression-engine: jinja
     executor:
       docker:
-        image: resolwebio/legacy:1.0.0
+        image: resolwebio/rnaseq:3.2.0
   data_name: "Converted GFF3 to GTF file"
-  version: 0.1.1
+  version: 0.2.0
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/gff_to_gtf.yml]: Let's make this 1.0.0 as version 1 signals production readyness.

File: resolwe_bio/processes/support_processors/gff_to_gtf.yml [hunk 14]
```diff
@@ -41,10 +49,24 @@
     runtime: polyglot
     language: bash
     program: |
-      NAME=`basename {{annotation.annot.file}} .gff`
-      gff3_to_gtf.py {{ annotation.annot.file }} "${NAME}.gtf"
 
-      re-save-file annot "${NAME}.gtf"
+      {% set name = ((annotation.annot.file|basename) | replace(".gff3","") | replace(".gff","")) %}
+      gff3_to_gtf.py {{ annotation.annot.file }} {{name}}.gtf
+      re-checkrc
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/gff_to_gtf.yml]: Include an error message.

File: resolwe_bio/processes/support_processors/gff_to_gtf.yml [hunk 15]
```diff
@@ -41,10 +49,24 @@
...
+
+      igvtools sort {{name}}.gtf {{name}}_sorted.gtf
+      re-checkrc
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/gff_to_gtf.yml]: Include an error message.

File: resolwe_bio/processes/support_processors/gff_to_gtf.yml [hunk 16]
```diff
@@ -41,10 +49,24 @@
...
+      igvtools index {{name}}_sorted.gtf
+      re-checkrc
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/gff_to_gtf.yml]: Include an error message.

File: resolwe_bio/processes/support_processors/gff_to_gtf.yml [hunk 17]
```diff
@@ -41,10 +49,24 @@
     runtime: polyglot
     language: bash
     program: |
-      NAME=`basename {{annotation.annot.file}} .gff`
-      gff3_to_gtf.py {{ annotation.annot.file }} "${NAME}.gtf"
 
-      re-save-file annot "${NAME}.gtf"
+      {% set name = ((annotation.annot.file|basename) | replace(".gff3","") | replace(".gff","")) %}
+      gff3_to_gtf.py {{ annotation.annot.file }} {{name}}.gtf
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/gff_to_gtf.yml]: Not sure if this and the next few calls work if the file name includes spaces. Please refactor the test to have space in the file name.

File: docs/CHANGELOG.rst [hunk 18]
```diff
@@ -40,6 +40,8 @@ Added
   - ``upload-gtf``
   - ``gff-to-gtf``
 - Add Tabix index for Jbrowse to ``upload-bed`` processor
+- Add BigWig, BigBed and JBrowse track outputs to ``macs14`` process
+- Add Species and Build outputs to ``rose2`` process
```
Comment (mstajdohar) [inline docs/CHANGELOG.rst]: Nice!

File: resolwe_bio/processes/chip_seq/rose2.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mstajdohar) [reply resolwe_bio/processes/chip_seq/rose2.yml]: How much data do we have? Can we simply delete and recompute?

File: resolwe_bio/processes/chip_seq/rose2.yml [hunk 19]
```diff
@@ -110,30 +116,59 @@
...
+      {% if input_upload and input_upload.species != rankby.species %}
+        re-error \
+          {{ ('Species of rankby bam file (%s) and upload bed file (%s) do not match. ' ~
+             'Please provide aligned reads and annotation with the same species.')
+             |format(input_upload.species, rankby.species) }}
+      {% endif %}
+
       {% if input_upload %}
-        NAME=`basename {{ input_upload.bed.file }} .bed`
+        if [[ {{ input_upload.bed.file }} == *.bed ]]; then
```
Comment (mstajdohar) [inline resolwe_bio/processes/chip_seq/rose2.yml]: I suggest we are consistent and use Jinja.

File: resolwe_bio/processes/chip_seq/macs2.yml [hunk 20]
```diff
@@ -183,6 +215,18 @@
         re-error "--broad can't be combined with --call-summits!"
       {% endif %}
 
+      {% if control and case.build != control.build %}
```
Comment (mstajdohar) [inline resolwe_bio/processes/chip_seq/macs2.yml]: Nice!

File: resolwe_bio/processes/chip_seq/scoring.yml [hunk 21]
```diff
@@ -33,7 +33,15 @@
     runtime: polyglot
     language: bash
     program: |
-      closest-features --dist {{ peaks.narrow_peaks.file }} {{ bed.bed.file }} > macs2_peakscore
+
+      if [[ {{ bed.bed.file }} == *.bed ]]
```
Comment (mstajdohar) [inline resolwe_bio/processes/chip_seq/scoring.yml]: I suggest to use Jinja.

File: resolwe_bio/processes/import_data/genome.yml [hunk 22]
```diff
@@ -143,10 +147,15 @@
       picard-tools CreateSequenceDictionary R="${FASTA_NAME}.fasta" O="${FASTA_NAME}.dict"
       re-checkrc
 
+      # Create JBrowse track
+      prepare-refseqs.pl --fasta "${FASTA_NAME}.fasta.gz" --out .
```
[the remark this answers is not in the document]
Comment (JenkoB) [reply resolwe_bio/processes/import_data/genome.yml]: I have tested it with spaces and it is working, however it is nicer if it is `"${FASTA_NAME}".fasta.gz`

File: docs/CHANGELOG.rst (hunk 12 above)
Comment (JenkoB) [reply docs/CHANGELOG.rst]: Why? These are three different processes.

File: resolwe_bio/processes/import_data/annotation.yml (hunk 11 above)
Comment (JenkoB) [reply resolwe_bio/processes/import_data/annotation.yml]: It is true, that in this case we wouldn't notice if it fails. I have added a test ``self.assertFields(upload_gtf, 'annot_sorted_idx_igv', {'file': 'Hs GRCh38_86 upload_sorted.gtf.idx', 'total_size': 116})``

**Comment by JenkoB** [comment]
I have applied all suggested changes, so please do another round of reviews.

File: resolwe_bio/processes/chip_seq/scoring.yml [hunk 23]
```diff
@@ -11,7 +11,7 @@
       docker:
         image: resolwebio/legacy:1.0.0
   data_name: "Peak score ({{ peaks.narrow_peaks.file|basename|default('?') }})"
-  version: 1.0.9
+  version: 1.0.10
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/chip_seq/scoring.yml]: Update to to newer docker image (`resolwebio/rnaseq`).

File: resolwe_bio/processes/chip_seq/rose2.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (JenkoB) [reply resolwe_bio/processes/chip_seq/rose2.yml]: On BCM server we have 129 rose2 data objects in different collections.
